File: src/dev_health_ops/api/utils/password_policy.py
from __future__ import annotations

from functools import lru_cache
from pathlib import Path

MIN_PASSWORD_LENGTH = 12
MAX_PASSWORD_LENGTH = 128


@lru_cache(maxsize=1)
def _load_common_passwords() -> set[str]:
    passwords_path = (
        Path(__file__).resolve().parents[2] / "data" / "common_passwords.txt"
    )
    if not passwords_path.exists():
        return set()

    with passwords_path.open("r", encoding="utf-8") as password_file:
        return {
            line.strip().lower()
            for line in password_file
            if line.strip() and not line.strip().startswith("#")
        }


def validate_password(password: str) -> list[str]:
    violations: list[str] = []

    password_length = len(password)
    if password_length < MIN_PASSWORD_LENGTH:
        violations.append(
            f"Password must be at least {MIN_PASSWORD_LENGTH} characters long"
        )
    if password_length > MAX_PASSWORD_LENGTH:
        violations.append(
            f"Password must be no more than {MAX_PASSWORD_LENGTH} characters long"
        )

    if not any(character.isalpha() for character in password):
        violations.append("Password must include at least one letter")
    if not any(character.isdigit() for character in password):
        violations.append("Password must include at least one number")

    if password.lower() in _load_common_passwords():
        violations.append("Password is too common")

    return violations

File: internal/auth/passwordpolicy/passwordpolicy.go
// Package passwordpolicy is password_policy.py: the one password strength
// check every route that sets a password applies (the admin set-password
// route and self-registration).
package passwordpolicy

import (
	_ "embed"
	"strings"
	"sync"
	"unicode"

	"github.com/full-chaos/dev-health-ops/internal/pythonparity"
)

//go:embed data/common_passwords.txt
var commonPasswordsFile string

const (
	minPasswordLength = 12
	maxPasswordLength = 128
)

var (
	commonPasswordsOnce sync.Once
	commonPasswords     map[string]struct{}
)

// loadCommonPasswords is password_policy.py's _load_common_passwords: one
// parse of the checked-in denylist, lowercased, comments (#) and blank
// lines dropped. The list is a byte-for-byte copy of
// src/dev_health_ops/data/common_passwords.txt, because `//go:embed` cannot
// reach outside the package directory; TestEmbeddedListMatchesThePythonOne
// fails when either side drifts.
func loadCommonPasswords() map[string]struct{} {
	commonPasswordsOnce.Do(func() {
		commonPasswords = map[string]struct{}{}
		for _, line := range strings.Split(commonPasswordsFile, "\n") {
			trimmed := strings.TrimSpace(line)
			if trimmed == "" || strings.HasPrefix(trimmed, "#") {
				continue
			}
			commonPasswords[strings.ToLower(trimmed)] = struct{}{}
		}
	})
	return commonPasswords
}

// Validate is password_policy.py's validate_password, same
// violation strings and same order (length checks, then letter, then
// digit, then common-password lookup).
func Validate(password string) []string {
	var violations []string

	length := len([]rune(password))
	if length < minPasswordLength {
		violations = append(violations, "Password must be at least 12 characters long")
	}
	if length > maxPasswordLength {
		violations = append(violations, "Password must be no more than 128 characters long")
	}

	// character.isalpha() has zero measured divergence from
	// unicode.IsLetter (both are Unicode category L*), so hasLetter needs
	// no parity helper. character.isdigit() is BROADER than
	// unicode.IsDigit (decimal-only): pythonparity.IsDigit is the one
	// canonical str.isdigit() -- e.g. "Abcdefghijk²" (U+00B2 SUPERSCRIPT
	// TWO) satisfies Python's digit requirement and unicode.IsDigit does
	// not.
	hasLetter, hasDigit := false, false
	for _, r := range password {
		if unicode.IsLetter(r) {
			hasLetter = true
		}
		if pythonparity.IsDigit(r) {
			hasDigit = true
		}
	}
	if !hasLetter {
		violations = append(violations, "Password must include at least one letter")
	}
	if !hasDigit {
		violations = append(violations, "Password must include at least one number")
	}

	if _, common := loadCommonPasswords()[strings.ToLower(password)]; common {
		violations = append(violations, "Password is too common")
	}

	return violations
}
